Contracts/feature/new solidity code from dom

josojo opened the pull request:
try do make the migration run and do the merge together

josojo committed:
new dutchExchange.sol

josojo committed:
new dutchExchange.sol

josojo committed:
new tulip and dutchexchange interface

josojo committed:
new tulip and dutchexchange interface

josojo committed:
it migrates, many tests are failing

josojo committed:
it migrates, many tests are failing

josojo committed:
editing to allow the function addtokenPair to run throught

josojo committed:
editing to allow the function addtokenPair to run throught

josojo committed:
oracle issue resolved. first test passes

josojo committed:
oracle issue resolved. first test passes

josojo committed:
test 1 passes, if we exclude tulip issuance code

josojo committed:
test 1 passes, if we exclude tulip issuance code

W3stside committed:
added eventWatcher and logger fns

W3stside committed:
added eventWatcher and logger fns

W3stside committed:
changed some c.logs to logger and formatting

W3stside committed:
changed some c.logs to logger and formatting

josojo committed:
second test passes, if we do not settle any fees

josojo committed:
second test passes, if we do not settle any fees

Velenir commented on contracts/DutchExchange.sol:
```diff
@@ -507,41 +494,39 @@ contract DutchExchange {
         require(sellerBalance > 0);
 
         // Get closing price for said auction
         fraction memory closingPrice = closingPrices[sellToken][buyToken][auctionIndex];
```
I think `closingPrice` here can be declared with `storage` instead of `memory`.
`memory` creates a copy of the closing price in contract's storage; `storage` -- only a reference.

Same on lines 633, 674, 786, 787

josojo replied:
We used memory because it stores the struct only temporarily and is cheaper to use than storage

Velenir committed:
changed api according to the latest contract version

Velenir committed:
made start_auction script work with the latest contract version

Velenir committed:
made get_auction_stats script work with the latest contract version

Velenir commented on contracts/DutchExchange.sol:
```diff
@@ -320,51 +303,58 @@ contract DutchExchange {
+
+            // C2
+            // R2.1: sell orders must go to next auction
+            if (getAuctionStart(sellToken, buyToken) < now && closingPriceDen > 0 && closingPriceDenOpp > 0){
```
Currently `getAuctionStart(sellToken, buyToken) < now` is always true at least for the first, freshly created toke pair, so I can repeatedly post sell orders to the ongoing auction.

josojo replied:
solved this one in a seperate commit

Velenir commented on contracts/DutchExchange.sol:
```diff
@@ -634,125 +628,87 @@ contract DutchExchange {
         public
         constant
-        returns (uint num, uint den)
+        returns (fraction price)
```
With these changes functions that return `structs` (`getPrice` and `priceOracle`) can't be called from JavaScript:

```
invalid solidity type!: tuple
```

`(uint num, uint den)`, which was returned previously, is properly converted into `BigNumber[2]`
It would be nice to see either return to the `(uint, uint)` convention or a helper function - wrapper over `getPrice`, something like:
```
function getPriceForJS(...) public constant returns (uint, uint) {
    fraction memory price = getPrice(...);
    return (price.num, price.den);
}
```

josojo replied:
ran into the same issues. Will write a wrapper function, since we need the returns fraction, to save variables.

josojo replied:
Got help from stefan and rimeissner:
this should be possible with the new solidity version
0.4.19
https://github.com/ethereum/solidity/blob/develop/Changelog.md#0419-2017-11-30

Velenir commented on contracts/DutchExchange.sol:
```diff
@@ -764,166 +720,200 @@ contract DutchExchange {
     function priceOracle(
         address token
+    )
+        public
```
This function is no longer `constant`, so can't be called from JavaScript to get a meaningful result, only transaction receipt. Is this change intentional? Also it returns a `struct`, same issue as above with `getPrice` function.

josojo replied:
made it constant again

Velenir committed:
changed sell_order script to work with the latest contract version

Velenir committed:
changed buy_order script to work with the latest contract version

Velenir committed:
moved price log higher so it doesn't depend on oracle prices being available

for now getPrice and priceOracle functions don't work from JS side due to contract changes

Velenir commented on contracts/DutchExchange.sol:
```diff
@@ -113,39 +108,41 @@ contract DutchExchange {
+    event AuctionCleared(
+        address sellToken,
+        address buyToken,
+        uint sellVoluem,
```
typo sellVolu~em~me

josojo replied:
solved in newest commit

Velenir committed:
changed getUnclaimedBuyerFunds to correspond to the new return signature

Velenir committed:
removed old get_auction_stats script

Velenir committed:
Merge branch 'contracts/feature/NewSolidityCodeFromDom' into trufflescripts2.1_for_DX-2.0

josojo committed:
solving venilr issue: getAuctionStart(sellToken, buyToken) < now is always true at least

josojo committed:
made priceOracle constant function, as requested from velenir

josojo committed:
corrected claimBuyerFunds

josojo committed:
2 tests are passing, no solidity code is commented out! fee things is running through

josojo committed:
newest smartcontract changes from dom incorporated

josojo committed:
newest smartcontract changes from dom incorporated

josojo committed:
finally all 3 tests are passing

josojo committed:
adds OWL functionallity to dutchExchange.js

josojo committed:
missing TokenOWL.sol

hacker-DOM committed:
Merge pull request #99 from gnosis/contracts/feature/AdditionOfOWLfunctionality

adds OWL functionallity to dutchExchange.sol

josojo committed:
error in TokenOWL corrected

josojo committed:
david found the error: standard tokens also need to be deployed for some reason

josojo committed:
addtion of withdrawfunction

W3stside committed:
changed weird pathing format in contracts + linted across files

W3stside committed:
changed weird pathing format in contracts + linted across files

josojo committed:
Merge pull request #103 from gnosis/contracts/feature/addtionOfTulipWithdraw

addtion of TULIP withdrawfunction

josojo committed:
added getPriceForJS and getPriceOracleForJS

W3stside committed:
refactored contract functions out into sep file

hacker-DOM committed:
tests behaving randomly

W3stside committed:
removed unnecessary async

W3stside committed:
linting

Velenir committed:
Merge pull request #91 from gnosis/trufflescripts2.1_for_DX-2.0

Trufflescripts 2.1 for NewSolidityCode PR

W3stside committed:
wait pckg works fine .. error in setAuction function was resulting in wrong auction time. Fixed

W3stside committed:
fixed most BULLSHIT bug ever concerning the INCORRECT auctionStart fn
--> dx.auctionStart SHOULD HAVE BEEN dx.getAuctionStart
--> Tests are PASSING x 4 ... yolo

W3stside committed:
Merge branch 'contracts/feature/NewSolidityCodeFromDom' into contracts/feature/David/RefactoringTestFunctions

W3stside committed:
Merge pull request #105 from gnosis/contracts/feature/David/RefactoringTestFunctions

refactored contract functions out into sep file - TESTS PASSING WOOHOOO

W3stside committed:
linting and fixed wrong tokenTUL name

Velenir committed:
implemented stopWatching function on eventWatcher utility function

Velenir committed:
stop watching events after tests are run

Velenir committed:
simplified stopWatching use in mocha test hooks

Velenir committed:
made stopWatching function distinguish between contracts

W3stside committed:
adding tulip tests

W3stside committed:
Merge pull request #104 from gnosis/contracts/feature/NewInterfaceForJSFunctions

added getPriceForJS and getPriceOracleForJS

Velenir committed:
simplified resetting event cache

Velenir committed:
fixed wrong variable used to mint Tulips

Velenir committed:
added assertions and fixed tulip  test

Velenir committed:
Merge pull request #106 from gnosis/tests/stopWatchingEvents

Tests/stop watching events

hacker-DOM committed:
Fee's in addTokenPair() should go to auctionIndex 2

hacker-DOM committed:
Merge branch 'contracts/feature/NewSolidityCodeFromDom' into contracts/tests/randomBehaviorOfTests

W3stside committed:
Merge pull request #108 from gnosis/contracts/tests/TulipIssuanceFix

Contracts/tests/tulip issuance fix

hacker-DOM committed:
Merge pull request #107 from gnosis/contracts/tests/randomBehaviorOfTests

Merge refactored code into "master" branch

hacker-DOM committed:
Changed require() to Log(); return;

W3stside committed:
adding simple unlocking locking testing for Tulips

W3stside committed:
Testing withdrawal time + global tulip state in 1 contract

hacker-DOM committed:
Fixed 'underflow' error, all tests passing except fuzzy

hacker-DOM committed:
Fixed buyOrders being accepted when auctionStart == 1

Velenir committed:
fixed linting errors and disabled a few rules where appropriate

Velenir committed:
consider transaction with Log event as rejected

Velenir committed:
replaced long if-else cascades with switch-case

await does work with switch()

Velenir committed:
mine a block after 'evm_increaseTime' as timestamp isn't changed otherwise

Velenir committed:
fixed linting issues

josojo committed:
fixes volnuerability to steal funds with only theorectical closed auctions

josojo committed:
makes reverted buyOrders cheaper, if auction is already closed due to another faster transactionn

josojo committed:
addition of further workflow tests, all tests are passing

Velenir committed:
added check for "R<number>" substring

W3stside committed:
all tests passing:
1. seller1 addNewTokPair --> correct sVol
2. buyer1 and buyer2 can make postBuy and claim
3. tulip accurately assigned
4. can lock and unlock 24 hours later
5. ???
6. Profit!
--> also moved fns into testFunctions.js

W3stside committed:
Merge branch 'contracts/DX-2.0-No-Frontend' of https://github.com/gnosis/dutch-exchange into testMerge/travis/contracts/feature/NewSolidityCodeFromDom

W3stside committed:
Merge pull request #123 from gnosis/travis/contracts/feature/NewSolidityCodeFromDom

Travis/contracts/feature/new solidity code from dom

W3stside committed:
adding my friend travis

Velenir committed:
Merge pull request #112 from gnosis/tests/fuzzy/assertRejects

Tests/fuzzy/assert rejects

hacker-DOM committed:
Merge pull request #115 from gnosis/contracts/fix/smallfixesOfSolidity

Updates to the contracts and further auctionflow tests

W3stside committed:
Merge branch 'contracts/feature/NewSolidityCodeFromDom' into contracts/tests/NewSolidityCodeFromDom/TulipTests

W3stside committed:
Merge pull request #110 from gnosis/contracts/tests/NewSolidityCodeFromDom/TulipTests

Tulip Token Testing
